WIP: Add pure python CLI for running SimulaQron Network

cgmcintyr opened the pull request:
This pull request is a work in progress, I'd like to get the authors' feedback.

The idea is to have a [Click](https://click.palletsprojects.com/en/7.x/) command for starting all nodes defined in the `$NETSIM/config` directory. See #100 for more details.

Python's `multiprocessing` module has been used to spawn virtual nodes using the main methods of `run/startNode.py` and `run/startCQC.py`.

Everything is working, however I am having a problem with CTRL-C and shutting down the virtual nodes gracefully. What is the current method of gracefully shutting down a running network? I can find a method for starting a `backEnd`, but no methods for stopping/terminating them.

cgmcintyr commented:
Currently:

* `run/startAll.sh` can be run using `./cli/main network start-all` and supports the same flags for generating config (`-nn`, `-tp`, etc.)
* `run/startAllLog.sh` can be run using `./cli/main network start-all-log` and supports the same flags for generating config (`-nn`, `-tp`, etc.)

AckslD commented:
Hi @cgmcintyr !

I have been very busy lately and unfortunately had no time to look at this. I hope to go over this and your other pull request in the coming days or next week :)

cgmcintyr commented:
Hi, I fixed the missing signal import

AckslD commented:
@cgmcintyr 

I like this work, I think it's probably a good idea to go away from starting things from shell-scripts.

There are still some things not working, for example reactor is not imported in `run/startNode.py`, `sys` is not imported in `cli/commands/network.py` etc. Furthermore, there's the option to set a Network simulation root (`--netsim`), however this is not used as far as I can tell. In  `run/startNode.py` `$NETSIM` is used.

For shutting down nodes previously we simply found the process IDs with grep and killed these. Not very nice so I think the approach you're taking is better.

cgmcintyr commented:
Apologies for the delay, work has been busy.

I fixed the missing imports, so it should be usable now :)

The `--netsim` flag is a feature not a bug, it's for the case where the user doesn't want to mess around with environment variables (in my experience a flag is easier for less experienced users to grasp). However if you would like it removed I can do that.

I changed the flags to single letters so `-nr` is now `-N`, `-nn` is now `-n`, and `-tp` is now `-t`. This is due to Click not being great at supporting multi-letter flags. The longer `--nodes`, `--nrnodes`, and `--topology` are still present.

Finally, you will notice on pressing `CTRL-C` after running a command like `./cli/main.py start-all -N 5` there are a lot of errors printed - basically I am not sure how to gracefully shut-down all the processes so these get shown to the user...

While doing this work, a thought occurred that this CLI has the role to run/monitor/shutdown a number of independent processes that can speak to each-other and bind to host network port. This sounds a perfect use case for containers. Have you considered having the `SimulaQron` package simply install scripts for running nodes, and then packaging these up in containers for end users to run as they want? The complexity of topologies/number of nodes/names/ip-addresses/etc could be handed off to the container management tool?

AckslD commented on cli/commands/network.py:
```diff
@@ -0,0 +1,160 @@
+from SimulaQron.settings import Settings
+from SimulaQron.virtNode.virtual import backEnd
+from SimulaQron.run.startCQC import main as start_cqc
+from SimulaQron.run.startCQCLog import main as start_cqc_log
```
Should be `SimulaQron.run.log.startCQCLog`

AckslD commented on cli/commands/network.py:
```diff
@@ -0,0 +1,160 @@
+
+@network.command()
+@common_options
+def start_all(netsim, nrnodes, nodes, topology):
```
Would it be possible to also add a "kill" command? For example if I start the network as a background process `./cli/main network start_all &` then it would be convenient to be able to kill it without having to find the process number manually.

AckslD commented on config/Nodes.cfg:
```diff
@@ -1,5 +1 @@
-Charlie
-David
-Eve
+1
```
Changes to these files shouldn't be included in the pull requests.

AckslD commented on config/appNodes.cfg:
```diff
@@ -5,8 +5,4 @@
-Charlie, localhost, 8813
-David, localhost, 8814
-Eve, localhost, 8815
+1, localhost, 8803
```
Changes to these files shouldn't be included in the pull requests.

AckslD commented on config/cqcNodes.cfg:
```diff
@@ -5,8 +5,4 @@
-Charlie, localhost, 8808
-David, localhost, 8809
-Eve, localhost, 8810
+1, localhost, 8802
```
Changes to these files shouldn't be included in the pull requests.

AckslD commented on config/virtualNodes.cfg:
```diff
@@ -5,8 +5,4 @@
-Charlie, localhost, 8803
-David, localhost, 8804
-Eve, localhost, 8805
+1, localhost, 8801
```
Changes to these files shouldn't be included in the pull requests.

AckslD commented on docs/ConfNodes.rst:
```diff
@@ -5,31 +5,35 @@ Configuring the simulated network
+The backend of a SimulaQron network is a set of running virtual nodes and their corresponding CQC servers. To start the backend of a SimulaQron network run the command ``./cli/main.py network start-all``.
 
-If you want to start a network with for example the three nodes Alex, Bart, Curt, simply type::
+With no arguments, a network is setup using the files in the ``$NETSIM/config`` directory. If no files exist in this directory, the ``./cli/main.py network start-all`` command will generate config files for five nodes (Alice, Bobm Charlie, David, and Eve) in a fully connected topology (every node is connected to every other node).
```
Bobm -> Bob

AckslD commented on docs/ConfNodes.rst:
```diff
@@ -5,31 +5,35 @@ Configuring the simulated network
 Which will start up a network with the nodes Alice, Bob, Charlie, Node0, Node1, ..., Node6. If --nrnodes is less than the entries in --nodes, then --nrnodes is ignored. The two keywords can also be specified shorter as -nd and -nn respectively. So the above can also be done as::
 
-    sh run/startAll.sh -nd "Alice Bob Charlie" -nn 10
+    ./cli/main.py network start-all -nd Alice,Bob,Charlie -nn 10
```
This should be `-n` now right?

AckslD commented on docs/ConfNodes.rst:
```diff
@@ -43,7 +47,7 @@ Along with setting up the network with the specified topology a .png figure is a
 As a final example let's combine all the arguments specified above and create a network using 15 nodes, where two of then are called Alice and Bob and the topology of the network is randomly generated as a connected graph with 20 edges::
 
-    sh run/startAll.sh -nd "Alice Bob" -nn 15 -tp random_connected_20
+    ./cli/main.py network start-all -nd Alice,Bob -nn 15 -tp random_connected_20
```
`-nn` -> `-N`, `-tp` -> `-t`

AckslD commented on run/startAllLog.sh:
```diff
@@ -11,16 +11,12 @@ then
-            names="$names $name"
-        done < "$NETSIM/config/Nodes.cfg"
-        python "$NETSIM/run/log/startCQCLog.py" $names &
+        python "$NETSIM/run/startCQCLog.py"
```
The path is wrong here `/run/log`

AckslD commented on run/startAllLog.sh:
```diff
@@ -11,16 +11,12 @@ then
 
         # We call this script again, without arguments, to use the newly created config-files
-        sh "$NETSIM/run/startAllLog.sh"
+        sh "$NETSIM/run/startAll.sh"
```
This should be startAllLog.sh

AckslD commented:
Hi @cgmcintyr !

I've continued working on the cli you started and created a new class (`simulaqron.network.Network`) which makes it easy to set up networks also within Python. I've also made all tests unittest and some other cleanup and reordering. See pull request #150 which also includes your commits. I therefore close this pull request. Feel free to add more features to the cli. For example it would be very nice if one can also use if for the settings. For example one might be able to write `./cli/SimulaQron.py set backend stabilizer` or similiar.

cgmcintyr commented:
Nice! Apologies I never got around to finishing it, glad my code was useful :)
